skynet/agent.py

```diff
@@ -41,3 +49,75 @@ async def create_tcpserver(port):
         await tcpserver.serve_forever()
     except KeyboardInterrupt:
         tcpserver.close()
+
+
+# Endpoints for the autoscaler to query the current state of the system
+class StateResponse(BaseModel):
+    request_status: str
+    current_state: str
+    desired_state: str
+
+
+class CurrentStateResponse(BaseModel):
+    current_state: str
+    active_connections: int
+    stress_level: float
+
+
+async def get_current_state():
+    try:
+        with open(STATE_FILE, 'r') as f:
+            return f.read().strip()
+    except FileNotFoundError:
+        log.error(f'The state file does not exist')
+        with open(STATE_FILE, 'w') as f:
+            f.write('ready')
+        return 'ready'
+    except Exception as e:
+        log.error(f'Failed to read the state file: {e}')
+        return 'unknown'
+
+
+@app.post('/state')
+async def echo_requests(request: Request):
+    global haproxy_check_state_task
+    payload = await request.json()
+    log.info(f'Received state change request: {payload}')
+    current_state = await get_current_state()
+
+    if haproxy_check_state_task is None:
+        loop = asyncio.get_running_loop()
+        haproxy_check_state_task = loop.create_task(update_haproxy_state())
+
+    if payload['state'] and payload['state'] in ['drain', 'maint', 'ready']:
+        with open(STATE_FILE, 'w') as f:
+            f.write(payload['state'].strip())
+        return StateResponse(request_status='success', current_state=current_state, desired_state=payload['state'])
+
+    log.error(f'Invalid state change request: {payload}')
+    return StateResponse(request_status='failed', current_state=current_state, desired_state=payload['state'])
+
+
+@app.get('/state')
+async def echo_requests(request: Request):
+    return CurrentStateResponse(
+        current_state=await get_current_state(),
+        active_connections=CONNECTIONS_METRIC._value.get(),
+        stress_level=TRANSCRIBE_STRESS_LEVEL_METRIC._value.get(),
+    )
+
+
+# Updates the HAProxy state if it changes via the API
+# so that the load balancer can react to the changes
+async def update_haproxy_state():
+    global haproxy_state
+    while True:
+        new_state = await get_current_state()
+        if new_state != haproxy_state:
+            log.info(f'HAProxy state changed from {haproxy_state} to {new_state}')
+            haproxy_state = new_state
+        if new_state in ('drain', 'maint'):
```

Comment:
Are we supposed to do anything in this state, such as reject new incoming requests?